### PR TITLE
Fixed bug where percentage discounts on order were causing +/- $0.01 …

### PR DESCRIPTION
**Pull request by johnmccombs1**
…differences

### CODE REVIEW

File: app/code/community/Bolt/Boltpay/Helper/Api.php [hunk 1]
```diff
@@ -316,6 +320,54 @@ protected function getRatesDebuggingData($rates) {
...
+            $difference = round(abs($boltDiscountAmount - $quoteDiscountAmount), 2);
+            if($difference <= 0.01) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    protected function fixQuoteDiscountAmount($transaction, $quote) {
```
Comment (daisy1754) [inline app/code/community/Bolt/Boltpay/Helper/Api.php]: Can you add comment (and maybe link to site describing magento's bug) to explain why this is needed and how it fixes the issue?

File: app/code/community/Bolt/Boltpay/Helper/Api.php [hunk 2]
```diff
@@ -316,6 +320,54 @@ protected function getRatesDebuggingData($rates) {
...
+        $quote->setTotalsCollectedFlag(false)->collectTotals()->save();
+
+        if($this->isDiscountRoundingDeltaError($transaction, $quote)) {
```
Comment (daisy1754) [inline app/code/community/Bolt/Boltpay/Helper/Api.php]: We should instead check `$boltDiscountAmount != $quoteDiscountAmount`

File: app/code/community/Bolt/Boltpay/Helper/Api.php [hunk 3]
```diff
@@ -316,6 +320,54 @@ protected function getRatesDebuggingData($rates) {
...
+                    $boltDiscountAmount -= $discount->amount->amount / 100;
+                }
+            }
+        }
+
+        return $boltDiscountAmount;
+    }
+
+    protected function isOnePennyRoundingError($boltDiscountAmount, $quoteDiscountAmount) {
+        if($boltDiscountAmount != $quoteDiscountAmount) {
```
Comment (daisy1754) [inline app/code/community/Bolt/Boltpay/Helper/Api.php]: are these float? What if `$boltDiscountAmount` is 10.00 and `$quoteDiscountAmount` is 10.000001 ?

File: app/code/community/Bolt/Boltpay/Helper/Api.php [hunk 4]
```diff
@@ -316,6 +320,54 @@ protected function getRatesDebuggingData($rates) {
...
+            $difference = round(abs($boltDiscountAmount - $quoteDiscountAmount), 2);
```
Comment (daisy1754) [inline app/code/community/Bolt/Boltpay/Helper/Api.php]: I prefer we don't assume sign of boltdiscountamount of quotediscountamount so `round(abs(A) - abs(B), 2)`

File: app/code/community/Bolt/Boltpay/Helper/Api.php (hunk 2 above)
Comment (johnmccombs1) [reply app/code/community/Bolt/Boltpay/Helper/Api.php]: That's essentially what this is doing, but it's checking to see if they are only off by $0.01. If they are off by more than this, then there is a different issue. Calling collectTotals could pull in other artifacts that could cause this number to be off by more than $0.01 (admin adds a new discount amount, etc) which would be a different error and would be caught by validateSubmittedOrder

File: app/code/community/Bolt/Boltpay/Helper/Api.php (hunk 3 above)
Comment (johnmccombs1) [reply app/code/community/Bolt/Boltpay/Helper/Api.php]: Both Bolt and Sales_Model_Service_Quote are rounded to two decimal places. In Magento collectTotals is always rounded just before returning the totals. If you think we should be more explicit we could round them to 2 decimal places here just to confirm. Thoughts?

File: app/code/community/Bolt/Boltpay/Helper/Api.php (hunk 4 above)
Comment (johnmccombs1) [reply app/code/community/Bolt/Boltpay/Helper/Api.php]: I considered this. But what about the case where the Bolt discount is positive and the Magento quote is negative (we're currently allowing this with shipping rate discounts). So for example, let's say we have a Bolt discount amount of $8.01 and a Magento discount amount of -$8.00. If we do round(abs(A) - abs(B), 2) we will end up with $8.01 - $8.00 = $0.01 which will incorrectly flag this error. Instead it should be $8.01 - (-$8.00) = $16.01 which is accurate and won't flag this as a rounding error.

File: app/code/community/Bolt/Boltpay/Helper/Api.php (hunk 1 above)
Comment (johnmccombs1) [reply app/code/community/Bolt/Boltpay/Helper/Api.php]: Pushed new change to address this.

**Pull request merged**